File: platforms/android/app/src/main/java/com/skillz/skillzcordova/SkillzCordova.java
package com.skillz.skillzcordova;

import android.app.Activity;

import org.apache.cordova.CordovaPlugin;
import org.apache.cordova.CallbackContext;

import com.skillz.Skillz;
import com.skillz.SkillzPlayer;
import com.skillz.model.MatchInfo;
import com.skillz.util.SkillzRandom;

import org.apache.cordova.CordovaWebView;
import org.apache.cordova.PluginResult;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.math.BigDecimal;
import java.util.HashMap;
import java.util.Iterator;
import java.util.Map;

public class SkillzCordova extends CordovaPlugin {
    private static CordovaWebView cordovaWebView = null;
    private static String onMatchWillBeginCallBackId = null;
    private static String onSkillzWillExitCallBackId = null;

    static final String LAUNCH_SKILLZ = "launchSkillz";
    static final String REPORT_SCORE = "reportScore";
    static final String ABORT_MATCH = "abortMatch";
    static final String IS_MATCH_IN_PROGRESS = "isMatchInProgress";
    static final String GET_RANDOM_NUMBER = "getRandomNumber";
    static final String GET_RANDOM_FLOAT = "getRandomFloat";
    static final String GET_RANDOM_NUMBER_WITH_RANGE = "getRandomNumberWithRange";
    static final String UPDATE_PLAYERS_CURRENT_SCORE = "updatePlayersCurrentScore";
    static final String GET_MATCH_RULES = "getMatchRules";
    static final String GET_MATCH_INFO = "getMatchInfo";
    static final String ADD_METADATA_FOR_MATCH_IN_PROGRESS = "addMetadataForMatchInProgress";
    static final String GET_SDK_VERSION = "getSDKVersion";
    static final String GET_CURRENT_PLAYER = "getCurrentPlayer";
    static final String SKILLZ_INIT = "skillzInit";
    static final String ON_MATCH_WILL_BEGIN = "callOnMatchWillBegin";
    static final String ON_SKILLZ_WILL_EXIT = "callOnSkillzWillExit";

    @Override
    public boolean execute(String action, JSONArray args, CallbackContext callbackContext) throws JSONException {
        Activity gameActivity = cordova.getActivity();
        switch(action) {
            case LAUNCH_SKILLZ:
                this.launchSkillz(gameActivity, callbackContext);
                return true;
            case REPORT_SCORE:
                this.reportScore(gameActivity, args, callbackContext);
                return true;
            case ABORT_MATCH:
                this.abortMatch(gameActivity, args, callbackContext);
                return true;
            case IS_MATCH_IN_PROGRESS:
                this.isMatchInProgress(callbackContext);
                return true;
            case GET_RANDOM_NUMBER:
                this.getRandomNumber(callbackContext);
                return true;
            case GET_RANDOM_FLOAT:
                this.getRandomFloat(callbackContext);
                return true;
            case GET_RANDOM_NUMBER_WITH_RANGE:
                this.getRandomNumberWithRange(args, callbackContext);
                return true;
            case UPDATE_PLAYERS_CURRENT_SCORE:
                this.updatePlayersCurrentScore(gameActivity, args, callbackContext);
                return true;
            case GET_MATCH_RULES:
                this.getMatchRules(callbackContext);
                return true;
            case GET_MATCH_INFO:
                this.getMatchInfo(gameActivity, callbackContext);
                return true;
            case ADD_METADATA_FOR_MATCH_IN_PROGRESS:
                this.addMetaDataForMatchInProgress(args.getString(0), args.getBoolean(1), callbackContext);
                return true;
            case GET_SDK_VERSION:
                this.getSDKVersion(callbackContext);
                return true;
            case GET_CURRENT_PLAYER:
                this.getCurrentPlayer(gameActivity, callbackContext);
                return true;
            case SKILLZ_INIT:
                this.skillzInit(gameActivity, args, callbackContext);
                return true;
            case ON_MATCH_WILL_BEGIN:
                this.setUpOnMatchWillBegin(callbackContext);
                return true;
            case ON_SKILLZ_WILL_EXIT:
                this.setUpOnSkillzWillExit(callbackContext);
                return true;
            default:
                return false;
        }
    }

    private void launchSkillz(Activity activity, CallbackContext callbackContext) {
        Skillz.launch(activity);
        callbackContext.success("Launched Skillz");
    }

    private void reportScore(Activity activity, JSONArray args, CallbackContext callbackContext) {
        try {
            BigDecimal bigDecScore = new BigDecimal(args.getDouble(0));
            String matchId = args.getString(1);
            Skillz.reportScore(activity, bigDecScore, matchId);
            callbackContext.success(bigDecScore.toString());
        } catch (JSONException e) {
            callbackContext.error("Missing score or match id argument");
        }
    }

    private void abortMatch(Activity activity, JSONArray args, CallbackContext callbackContext) {
        try {
            String matchId = args.getString(0);
            Skillz.abortMatch(activity, matchId);
            callbackContext.success("Skillz Match aborted");
        } catch (JSONException e) {
            callbackContext.error("Expected match id argument.");
        }
    }

    private void isMatchInProgress(CallbackContext callbackContext) {
        int isMatchInProgress = Skillz.isMatchInProgress() ? 1 : 0;
        callbackContext.success(isMatchInProgress);
    }

    private void getRandomNumber(CallbackContext callbackContext) {
        try {
            SkillzRandom rand = new SkillzRandom();
            JSONObject response = new JSONObject();
            response.put("value", rand.nextInt());
            callbackContext.success(response);
        } catch (JSONException e) {
            e.printStackTrace();
            callbackContext.error(e.getMessage());
        }
    }

    private void getRandomFloat(CallbackContext callbackContext) {
        try {
            SkillzRandom rand = new SkillzRandom();
            JSONObject response = new JSONObject();
            response.put("value", rand.nextFloat());
            callbackContext.success(response);
        } catch (JSONException e) {
            e.printStackTrace();
            callbackContext.error(e.getMessage());
        }
    }

    private void getRandomNumberWithRange(JSONArray args, CallbackContext callbackContext) {
        if (args.length() < 2) {
            callbackContext.error("Not enough arguments");
        } else {
            try {
                float min = (float) args.getDouble(0);
                float max = (float) args.getDouble(1);
                SkillzRandom rand = new SkillzRandom();
                float value = min + (rand.nextFloat() * (max - min));
                JSONObject response = new JSONObject();
                response.put("value", value);
                callbackContext.success(response);
            } catch(JSONException e) {
                e.printStackTrace();
                callbackContext.error(e.getMessage());
            }
        }
    }

    private void updatePlayersCurrentScore(Activity activity, JSONArray args, CallbackContext callbackContext) {
        try {
            BigDecimal bigDecScore = new BigDecimal(args.getDouble(0));
            String matchId = args.getString(1);
            Skillz.updatePlayersCurrentScore(activity, bigDecScore, matchId);
            callbackContext.success(bigDecScore.toString());
        } catch (JSONException e) {
            callbackContext.error("Expected match id argument.");
        }
    }

    private void getMatchRules(CallbackContext callbackContext) {
        Map<String, String> rules = Skillz.getMatchRules();
        JSONObject jsonFormattedRules = new JSONObject(rules);
        callbackContext.success(jsonFormattedRules);
    }

    private JSONObject matchToJson(MatchInfo match) throws JSONException {
        JSONObject object = new JSONObject();
        object.put("matchDescription", match.getMatchDescription());
        object.put("entryCash", match.getEntryCash());
        object.put("entryPoints", match.getEntryPoints());
        object.put("id", match.getId());
        object.put("templateId", match.getTemplateId());
        object.put("name", match.getName());
        object.put("isCash", match.isCash());
        object.put("isSynchronous", match.isSynchronous());
        object.put("skillzDifficulty", match.getSkillzDifficulty());
        object.put("players", match.getPlayers());
        return object;
    }

    private void getMatchInfo(Activity activity, CallbackContext callbackContext) {
        try {
            MatchInfo match = Skillz.getMatchInfo(activity);
            JSONObject jsonFormattedMatch = this.matchToJson(match);
            callbackContext.success(jsonFormattedMatch);
        } catch (JSONException e) {
            e.printStackTrace();
            callbackContext.error(e.getMessage());
        }
    }

    private void addMetaDataForMatchInProgress(String data, boolean isMatchInProgress, CallbackContext callbackContext) {
        try {
            HashMap<String, Object> attributes = new HashMap<>();
            JSONObject jsonData = new JSONObject(data);
            Iterator<String> keys = jsonData.keys();

            while (keys.hasNext()) {
                String key = keys.next();
                Object value = jsonData.get(key);

                attributes.put(key, value);
            }
            Skillz.addMetadataForMatchInProgress(attributes, isMatchInProgress);
            callbackContext.success("Skillz meta data added");
        } catch (JSONException e) {
            e.printStackTrace();
            callbackContext.error("Invalid format for meta data");
        }
    }

    private void getSDKVersion(CallbackContext callbackContext) {
        String sdkVersion = Skillz.SDKVersionShort();
        if (sdkVersion != null) {
            callbackContext.success(sdkVersion);
        } else {
            callbackContext.error("sdk version not found");
        }
    }

    private void getCurrentPlayer(Activity activity, CallbackContext callbackContext) {
        SkillzPlayer player = Skillz.getPlayerDetails(activity);
        try {
            JSONObject playerObject = new JSONObject();
            playerObject.put("id", player.getUserId());
            playerObject.put("displayName", player.getUserName());
            playerObject.put("avatarURL", player.getAvatarUrl());
            playerObject.put("flagURL", player.getFlagUrl());
            playerObject.put("playerMatchId", player.getPlayerMatchId());
            playerObject.put("isCurrentPlayer", player.isCurrentPlayer());
            callbackContext.success(playerObject);
        } catch (JSONException e) {
            callbackContext.error(e.getMessage());
        }
    }

    private void skillzInit(Activity activity, JSONArray args, CallbackContext callbackContext) {
        if (args.length() < 2) {
            callbackContext.error("Missing gameId or environment argument");
            return;
        }

        try {
            String gameId = args.getString(0);
            String environment = args.getString(1);
            String isProduction;
            if (environment.equals("SkillzSandbox")) {
                isProduction = "false";
            } else if (environment.equals("SkillzProduction")) {
                isProduction = "true";
            } else {
                callbackContext.error("Wrong environment argument. Use SkillzSandbox or SkillzProduction");
                return;
            }
            Skillz.setCrossplatformGameId(activity, gameId);
            Skillz.setCrossplatformGameEnvironment(activity, isProduction);
            Skillz.setCordovaDelegate(activity, this);
            callbackContext.success("Skillz Initialized");
        } catch(Exception e) {
            callbackContext.error(e.getMessage());
        }
    }

    private void setStaticCordovaWebView() {
        if (SkillzCordova.cordovaWebView == null) {
            SkillzCordova.cordovaWebView = this.webView;
        }
    }

    private void setUpOnMatchWillBegin(CallbackContext callbackContext) {
        setStaticCordovaWebView();
        SkillzCordova.onMatchWillBeginCallBackId = callbackContext.getCallbackId();
    }

    private void setUpOnSkillzWillExit(CallbackContext callbackContext) {
        setStaticCordovaWebView();
        SkillzCordova.onSkillzWillExitCallBackId = callbackContext.getCallbackId();
    }

    public static void onMatchWillBegin(String matchId) {
        if (SkillzCordova.onMatchWillBeginCallBackId != null) {
            PluginResult result = new PluginResult(PluginResult.Status.OK, matchId);
            result.setKeepCallback(true);
            SkillzCordova.cordovaWebView.sendPluginResult(result, SkillzCordova.onMatchWillBeginCallBackId);
        }
    }

    public static void onSkillzWillExit() {
        if (SkillzCordova.onSkillzWillExitCallBackId != null) {
            PluginResult result = new PluginResult(PluginResult.Status.OK);
            result.setKeepCallback(true);
            SkillzCordova.cordovaWebView.sendPluginResult(result, SkillzCordova.onSkillzWillExitCallBackId);
        }
    }
}

File: platforms/ios/Cordova 2048/Plugins/skillz-crossplatform-cordova/SkillzCordova.m
//
//  SkillzCordova.m
//  SkillzCordova
//
//  Created by Gabriel Firmacion on 8/7/18.
//
//
#import "SkillzCordova.h"
#import <Skillz/Skillz.h>

@implementation SkillzCordova

- (void)tournamentWillBegin:(NSDictionary *)gameParameters
              withMatchInfo:(SKZMatchInfo *)matchInfo
{
    #pragma GCC diagnostic push
    #pragma GCC diagnostic ignored "-Wundeclared-selector"
    NSString *gameParametersJsonString = [matchInfo performSelector:@selector(JSONStringRepresentation:)
                                                         withObject:gameParameters];

    #pragma GCC diagnostic pop
    NSString* jsString = [NSString stringWithFormat:@"SkillzDelegateCordova.onMatchWillBegin('%@');", gameParametersJsonString];
    [self.commandDelegate evalJs:jsString];
}

- (void)skillzWillExit
{
    [self.commandDelegate evalJs:@"SkillzDelegateCordova.onSkillzWillExit();"];
}

- (void)skillzInit:(CDVInvokedUrlCommand*) command
{
    NSArray *commandParams = command.arguments;
    if ([commandParams count] < 2) {
      CDVPluginResult *result = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"Missing gameId or environment argument"];
      [self.commandDelegate sendPluginResult:result callbackId:command.callbackId];
      return;
    }

    NSString *gameId = [command.arguments objectAtIndex:0];
    if ([gameId isKindOfClass:[NSNull class]]) {
        CDVPluginResult *result = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"Expected gameId argument."];
        [self.commandDelegate sendPluginResult:result callbackId:command.callbackId];
        return;
    }

    NSString *environment = [commandParams objectAtIndex:1];
    if ([environment isKindOfClass:[NSNull class]]) {
        CDVPluginResult *result = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"Expected environment argument."];
        [self.commandDelegate sendPluginResult:result callbackId:command.callbackId];
        return;
    }
    SkillzEnvironment skillzEnvironment;
    if ([environment isEqualToString:@"SkillzSandbox"]) {
        skillzEnvironment = SkillzSandbox;
    } else if ([environment isEqualToString:@"SkillzProduction"]) {
        skillzEnvironment = SkillzProduction;
    } else {
        CDVPluginResult *result = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"Wrong environment argument. Use SkillzSandbox or SkillzProduction"];
        [self.commandDelegate sendPluginResult:result callbackId:command.callbackId];
        return;
    }

    [[Skillz skillzInstance] initWithGameId:gameId
                            forDelegate:self
                        withEnvironment:skillzEnvironment
                              allowExit:YES];
    CDVPluginResult *result = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsString:@"Skillz Initialized"];
    [self.commandDelegate sendPluginResult:result callbackId:command.callbackId];
}

- (void)launchSkillz:(CDVInvokedUrlCommand*) command
{
    [[Skillz skillzInstance] launchSkillz];
    CDVPluginResult *result = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsString:@"Launched Skillz"];
    [self.commandDelegate sendPluginResult:result callbackId:command.callbackId];
}

- (void)isMatchInProgress:(CDVInvokedUrlCommand*)command
{
    BOOL istournamentinProgress = [[Skillz skillzInstance] tournamentIsInProgress];
    int isTournamentInProgressInt = istournamentinProgress ? 1 : 0;
    CDVPluginResult *result = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsInt:isTournamentInProgressInt];
    [self.commandDelegate sendPluginResult:result callbackId:command.callbackId];
}

- (void)reportScore:(CDVInvokedUrlCommand*)command
{
    NSArray *commandParams = command.arguments;
    if ([commandParams count] < 2) {
      CDVPluginResult *result = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"Missing score or match id argument"];
      [self.commandDelegate sendPluginResult:result callbackId:command.callbackId];
      return;
    }

    NSNumber *score = [commandParams objectAtIndex:0];
    if ([score isKindOfClass:[NSNull class]]) {
        CDVPluginResult *result = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"Expected score argument."];
        [self.commandDelegate sendPluginResult:result callbackId:command.callbackId];
        return;
    }

    NSNumber *matchId = [commandParams objectAtIndex:1];
    if ([matchId isKindOfClass:[NSNull class]]) {
        CDVPluginResult *result = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"Expected match id argument."];
        [self.commandDelegate sendPluginResult:result callbackId:command.callbackId];
        return;
    }

    [[Skillz skillzInstance] displayTournamentResultsWithScore:score
                                                   withMatchId:matchId
                                                withCompletion:^{
                                                    // Code in this block is called when exiting to Skillz
                                                    // and reporting the score.You may need to call a JS function that resets the state of your game.
                                                    // e.g.
                                                    // NSString* jsString = [NSString stringWithFormat:@"resetGame(\"%@\");", @""];
                                                    // [self.webView stringByEvaluatingJavaScriptFromString:jsString];
                                                }];
    CDVPluginResult *result = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsString:@"Reported Score"];
    [self.commandDelegate sendPluginResult:result callbackId:command.callbackId];
}

- (void)updatePlayersCurrentScore:(CDVInvokedUrlCommand*)command
{
    NSArray *commandParams = command.arguments;
    if ([commandParams count] < 2) {
      CDVPluginResult *result = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"Missing score or match id argument"];
      [self.commandDelegate sendPluginResult:result callbackId:command.callbackId];
      return;
    }

    NSNumber *score = [commandParams objectAtIndex:0];
    if ([score isKindOfClass:[NSNull class]]) {
        CDVPluginResult *result = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"Expected score argument."];
        [self.commandDelegate sendPluginResult:result callbackId:command.callbackId];
        return;
    }

    NSNumber *matchId = [commandParams objectAtIndex:1];
    if ([matchId isKindOfClass:[NSNull class]]) {
        CDVPluginResult *result = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"Expected match id argument."];
        [self.commandDelegate sendPluginResult:result callbackId:command.callbackId];
        return;
    }

    [[Skillz skillzInstance] updatePlayersCurrentScore:score
                                           withMatchId:matchId];
    CDVPluginResult *result = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsString:@"Updated Score"];
    [self.commandDelegate sendPluginResult:result callbackId:command.callbackId];
}

- (void)abortMatch:(CDVInvokedUrlCommand*)command
{
    NSArray *commandParams = command.arguments;
    if ([commandParams count] < 1) {
      CDVPluginResult *result = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"Missing match id argument"];
      [self.commandDelegate sendPluginResult:result callbackId:command.callbackId];
      return;
    }

    NSNumber *matchId = [commandParams objectAtIndex:0];
    if ([matchId isKindOfClass:[NSNull class]]) {
        CDVPluginResult *result = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"Expected match id argument."];
        [self.commandDelegate sendPluginResult:result callbackId:command.callbackId];
        return;
    }

    [[Skillz skillzInstance] notifyPlayerAbortWithMatchId:matchId
                                           WithCompletion:^{
      // Code in this block is called when exiting to Skillz
      // and aborting the match.You may need to call a JS function that resets the state of your game.
      // e.g.
      // NSString* jsString = [NSString stringWithFormat:@"resetGame(\"%@\");", @""];
      // [self.webView stringByEvaluatingJavaScriptFromString:jsString];
    }];
    CDVPluginResult *result = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsString:@"Skillz Match aborted"];
    [self.commandDelegate sendPluginResult:result callbackId:command.callbackId];
}

- (void)getMatchRules:(CDVInvokedUrlCommand*)command
{
    NSDictionary *matchRules = [Skillz getMatchRules];
    CDVPluginResult *result = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsDictionary:matchRules];
    [self.commandDelegate sendPluginResult:result callbackId:command.callbackId];
}

- (void)getMatchInfo:(CDVInvokedUrlCommand*)command
{
    SKZMatchInfo *matchInfo = [[Skillz skillzInstance] getMatchInfo];
    NSArray *players = [self convertPlayersToReadableObjects:matchInfo.players];
    NSMutableDictionary *matchInfoDict = [[NSMutableDictionary alloc]initWithCapacity:9];
    if (![[NSNumber numberWithInt:(int)matchInfo.id] isKindOfClass:[NSNull class]]) {
      [matchInfoDict setObject:[NSNumber numberWithInt:(int)matchInfo.id] forKey:@"id"];
    }

    if (![matchInfo.matchDescription isKindOfClass:[NSNull class]]) {
      [matchInfoDict setObject:matchInfo.matchDescription forKey:@"matchDescription"];
    }

    if (![matchInfo.entryCash isKindOfClass:[NSNull class]]) {
      [matchInfoDict setObject:matchInfo.entryCash forKey:@"entryCash"];
    }

    if (![matchInfo.entryPoints isKindOfClass:[NSNull class]]) {
      [matchInfoDict setObject:matchInfo.entryPoints forKey:@"entryPoints"];
    }

    if (![[NSNumber numberWithBool:matchInfo.isCash]  isKindOfClass:[NSNull class]]) {
      [matchInfoDict setObject:[NSNumber numberWithBool:matchInfo.isCash] forKey:@"isCash"];
    }

    if (![[NSNumber numberWithBool:matchInfo.isSynchronous] isKindOfClass:[NSNull class]]) {
      [matchInfoDict setObject:[NSNumber numberWithBool:matchInfo.isSynchronous] forKey:@"isSynchronous"];
    }

    if (![matchInfo.name isKindOfClass:[NSNull class]]) {
      [matchInfoDict setObject:matchInfo.name forKey:@"name"];
    }

    if (![players isKindOfClass:[NSNull class]]) {
      [matchInfoDict setObject:players forKey:@"players"];
    }

    if (![matchInfo.templateId isKindOfClass:[NSNull class]]) {
      [matchInfoDict setObject:matchInfo.templateId forKey:@"templateId"];
    }

    CDVPluginResult *result = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsDictionary:matchInfoDict];
    [self.commandDelegate sendPluginResult:result callbackId:command.callbackId];
}

- (NSArray*)convertPlayersToReadableObjects:(NSArray*)players
{
    NSMutableArray *playersArray = [NSMutableArray array];
    for (id player in players) {
      SKZPlayer *skzplayer = player;
      NSMutableDictionary *playerInfoDict = [[NSMutableDictionary alloc]initWithCapacity:6];

      if (![skzplayer.id isKindOfClass:[NSNull class]]) {
        [playerInfoDict setObject:skzplayer.id forKey:@"id"];
      }

      if (![skzplayer.playerMatchId isKindOfClass:[NSNull class]]) {
        [playerInfoDict setObject:skzplayer.playerMatchId forKey:@"playerMatchId"];
      }

      if (![[NSNumber numberWithBool:skzplayer.isCurrentPlayer] isKindOfClass:[NSNull class]]) {
        [playerInfoDict setObject:[NSNumber numberWithBool:skzplayer.isCurrentPlayer] forKey:@"isCurrentPlayer"];
      }

      if (![skzplayer.avatarURL isKindOfClass:[NSNull class]]) {
        [playerInfoDict setObject:skzplayer.avatarURL forKey:@"avatarURL"];
      }

      if (![skzplayer.displayName isKindOfClass:[NSNull class]]) {
        [playerInfoDict setObject:skzplayer.displayName forKey:@"displayName"];
      }

      if (![skzplayer.flagURL isKindOfClass:[NSNull class]]) {
        [playerInfoDict setObject:skzplayer.flagURL forKey:@"flagURL"];
      }

      [playersArray addObject:playerInfoDict];
    }

    return playersArray;
}

- (void)getCurrentPlayer:(CDVInvokedUrlCommand*)command
{
    SKZPlayer *currentPlayer = [Skillz player];
    NSMutableDictionary *currentPlayerDict = [[NSMutableDictionary alloc]initWithCapacity:6];
    if (![currentPlayer.id isKindOfClass:[NSNull class]]) {
      [currentPlayerDict setObject:currentPlayer.id forKey:@"id"];
    }

    if (![currentPlayer.playerMatchId isKindOfClass:[NSNull class]]) {
      SKZMatchInfo *matchInfo = [[Skillz skillzInstance] getMatchInfo];
      NSArray *players = [self convertPlayersToReadableObjects:matchInfo.players];
        NSString *currentPlayerMatchId;
        for (NSObject* player in players)
        {
            if ([player valueForKey:@"isCurrentPlayer"]) {
                currentPlayerMatchId = [NSString stringWithFormat:@"%@",[player valueForKey:@"playerMatchId"]];
            }
        }
      [currentPlayerDict setObject:currentPlayerMatchId forKey:@"playerMatchId"];
    }

    if (![[NSNumber numberWithBool:currentPlayer.isCurrentPlayer] isKindOfClass:[NSNull class]]) {
      [currentPlayerDict setObject:[NSNumber numberWithBool:currentPlayer.isCurrentPlayer] forKey:@"isCurrentPlayer"];
    }

    if (![currentPlayer.avatarURL isKindOfClass:[NSNull class]]) {
      [currentPlayerDict setObject:currentPlayer.avatarURL forKey:@"avatarURL"];
    }

    if (![currentPlayer.displayName isKindOfClass:[NSNull class]]) {
      [currentPlayerDict setObject:currentPlayer.displayName forKey:@"displayName"];
    }

    if (![currentPlayer.flagURL isKindOfClass:[NSNull class]]) {
      [currentPlayerDict setObject:currentPlayer.flagURL forKey:@"flagURL"];
    }

    CDVPluginResult *result = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsDictionary:currentPlayerDict];
    [self.commandDelegate sendPluginResult:result callbackId:command.callbackId];
}

- (void)addMetaDataForMatchInProgress:(CDVInvokedUrlCommand*)command
{
    NSString *metaDataString = [command.arguments objectAtIndex:0];
    BOOL isMatchInProgress = [command.arguments objectAtIndex:1];

    if ([metaDataString isKindOfClass:[NSNull class]]) {
      CDVPluginResult *result = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"Invalid format for meta data"];
      [self.commandDelegate sendPluginResult:result callbackId:command.callbackId];
      return;
    }

    NSError *jsonError;
    NSData *objectData = [metaDataString dataUsingEncoding:NSUTF8StringEncoding];
    NSDictionary *metaDataDict = [NSJSONSerialization JSONObjectWithData:objectData
                                                                 options:NSJSONReadingMutableContainers
                                                                   error:&jsonError];

    [[Skillz skillzInstance] addMetadata:metaDataDict forMatchInProgress:isMatchInProgress];
    CDVPluginResult *result = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsString:@"Added meta data"];
    [self.commandDelegate sendPluginResult:result callbackId:command.callbackId];
}

- (void)getSDKVersion:(CDVInvokedUrlCommand*)command
{
    NSString *sdkVersion = [Skillz SDKShortVersion];
    if ([sdkVersion isKindOfClass:[NSNull class]]) {
      CDVPluginResult *result = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"sdk version not found"];
      [self.commandDelegate sendPluginResult:result callbackId:command.callbackId];
      return;
    }
    CDVPluginResult *result = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsString:sdkVersion];
    [self.commandDelegate sendPluginResult:result callbackId:command.callbackId];
}

- (void)getRandomNumber:(CDVInvokedUrlCommand *)command
{
    NSMutableDictionary *randomNumDict = [[NSMutableDictionary alloc]initWithCapacity:1];
    [randomNumDict setObject:[NSNumber numberWithInt:(int)[Skillz getRandomNumber]] forKey:@"value"];
    if ([[randomNumDict valueForKey:@"value"] isKindOfClass:[NSNull class]]) {
      CDVPluginResult *result = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"getRandomNumber failed"];
      [self.commandDelegate sendPluginResult:result callbackId:command.callbackId];
      return;
    }
    CDVPluginResult *result = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsDictionary:randomNumDict];
    [self.commandDelegate sendPluginResult:result callbackId:command.callbackId];
}

- (void)getRandomFloat:(CDVInvokedUrlCommand *)command
{
    NSMutableDictionary *randomFloatDict = [[NSMutableDictionary alloc]initWithCapacity:1];
    [randomFloatDict setObject:[NSNumber numberWithFloat:(float)[Skillz getRandomFloat]] forKey:@"value"];
    if ([[randomFloatDict valueForKey:@"value"] isKindOfClass:[NSNull class]]) {
      CDVPluginResult *result = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"getRandomFloat failed"];
      [self.commandDelegate sendPluginResult:result callbackId:command.callbackId];
      return;
    }
    CDVPluginResult *result = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsDictionary:randomFloatDict];
    [self.commandDelegate sendPluginResult:result callbackId:command.callbackId];
}

- (void)getRandomNumberWithRange:(CDVInvokedUrlCommand *)command
{
    NSUInteger min = [[command.arguments objectAtIndex:0] integerValue];
    NSUInteger max = [[command.arguments objectAtIndex:1] integerValue];
    NSUInteger randomNum = [Skillz getRandomNumberWithMin:min andMax:max];
    NSMutableDictionary *randomNumDict = [[NSMutableDictionary alloc]initWithCapacity:1];
    [randomNumDict setObject:[NSNumber numberWithUnsignedInteger:randomNum] forKey:@"value"];
    if ([[randomNumDict valueForKey:@"value"] isKindOfClass:[NSNull class]]) {
      CDVPluginResult *result = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"getRandomNumberWithRange failed"];
      [self.commandDelegate sendPluginResult:result callbackId:command.callbackId];
      return;
    }
    CDVPluginResult *result = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsDictionary:randomNumDict];
    [self.commandDelegate sendPluginResult:result callbackId:command.callbackId];
}
@end
